Thermochimica API support for MQMQA (#24553)
idaholab/moose: modules/chemical_reactions/doc/content/source/actions/ChemicalCompositionAction.md

> The `ChemicalComposition` action is used to initiate the framework for thermochemical calculations using the thermodynamics solver [_Thermochimica_](https://github.com/ORNL-CEES/thermochimica). The action creates variables needed for the analysis, reads the thermodynamic model for material system, and sets the units for temperature, pressure, and elemental amounts. It can optionally initiate the amounts of elements used in the model.

Reviewer: ```suggestion
The `ChemicalComposition` action is used to initiate the framework for thermochemical calculations using the thermodynamics solver [*Thermochimica*](https://github.com/ORNL-CEES/thermochimica). The action creates variables needed for the analysis, reads the thermodynamic model for material system, and sets the units for temperature, pressure, and elemental amounts. It can optionally initiate the amounts of elements used in the model.
```

> The `ChemicalComposition` action is also used to specify lists of chemical phases and chemical species for which concentrations should be output at each timestep. The phase and species names must match those in the data file specified. The `ChemicalComposition` action creates all AuxVariables needed for output of phase concentration data, chemical potentials, vapor pressures, etc. The list of phases is specified in the `output_phases` variable, and the list of species is specified in `output_species`. The species must be specified in the convention `phase_name:species_name` so that the desired species in the desired phase is uniquely identified. Alternatively, users can specify `ALL` to output all species present in the specified thermodynamic database. Other desired outputs such as element potentials, vapor pressures of species in gas phase and moles of elements in specified phases can be specified using `output_element_potentials`, `output_vapor_pressures`, and `output_element_phases` respectively. The `ChemicalComposition` action creates a UserObject derived from `ThermochimicaNodalData` which runs Thermochimica at each node and stores the output in the created aux variables.

Reviewer: ThermochimicaNodalData should link to ThermochimicaNodalData.md

The `ChemicalComposition` action is used to initiate the framework for thermochemical calculations using the thermodynamics solver [_Thermochimica_](https://github.com/ORNL-CEES/thermochimica). The action creates variables needed for the analysis, reads the thermodynamic model for material system, and sets the units for temperature, pressure, and elemental amounts. It can optionally initiate the amounts of elements used in the model.

> The `ChemicalComposition` action is also used to specify lists of chemical phases and chemical species for which concentrations should be output at each timestep. The phase and species names must match those in the data file specified. The `ChemicalComposition` action creates all AuxVariables needed for output of phase concentration data. The list of phases is specified in the `output_phases` variable, and the list of species is specified in `output_species`. The species must be specified in the convention `phase_name:species_name` so that the desired species in the desired phase is uniquely identified. Alternatively, users can specify `ALL` to output all species. Other desired outputs such as element potentials, vapor pressures of species in gas phase and moles of elements in specified phases can be specified using `output_element_potentials`, `output_vapor_pressures`, and `output_element_phases` respectively. The `ChemicalComposition` action creates a UserObject derived from `ThermochimicaNodalData` which runs Thermochimica at each node and stores the output variables.

Reviewer: ```suggestion
The `ChemicalComposition` action is also used to specify lists of chemical phases and chemical species for which concentrations should be output at each timestep. The phase and species names must match those in the data file specified. The `ChemicalComposition` action creates all AuxVariables needed for output of phase concentration data, chemical potentials, vapor pressures, etc.. The list of phases is specified in the `output_phases` variable, and the list of species is specified in `output_species`. The species must be specified in the convention `phase_name:species_name` so that the desired species in the desired phase is uniquely identified. Alternatively, users can specify `ALL` to output all species. Other desired outputs such as element potentials, vapor pressures of species in gas phase and moles of elements in specified phases can be specified using `output_element_potentials`, `output_vapor_pressures`, and `output_element_phases` respectively. The `ChemicalComposition` action creates a UserObject derived from `ThermochimicaNodalData` which runs Thermochimica at each node and stores the output in the created auxvariables.
```

!listing id=chemact:globpar caption=Chemical elements to be used in the model
[GlobalParams]
  elements = 'Mo Ru'
[]

The vector `elements` will be passed to the `ChemicalComposition` action and used to generate scalar variables, (e.g. `Mo`, and `Ru`) of the chemical elements contained in the vector.

!alert note
The names of scalar variables listed in `elements` vector and created using `ChemicalComposition` action should not be used for other variables in the model, e.g. the model should not define variables `Mo`, or `Ru` in the `[Variables]` or `[AuxVariables]` blocks.

The element composition scalar variables can be initialized in `ChemicalComposition` action, or using `[Functions]` block combined with other MOOSE capabilities.

## Variables for Chemical Phases, Species, and Element Chemical Potentials

Chemical phase and species variables, as well as chemical potentials of elements, to be used and initiated in the model using in `ChemicalComposition` action are specified using the block `[GlobalParams]`, and vector `output_phases` as in [phaseact:globpar]:

> !listing id=phaseact:globpar caption=Chemical phases species to be used in the model, and list of desired output variables.

Reviewer: ```suggestion
!listing id=phaseact:globpar start=GlobalParams end=[] include-end=True caption=Chemical phases species to be used in the model, and list of desired output variables.
```
